databuilder/utils/date_utils.py

```diff
@@ -52,17 +53,17 @@ def date_add_months(date, num_months):
         return datetime.date(new_year, new_month, date.day)
     except ValueError:
         # We should only ever get an error for:
-        # 1) An out-of-range year, if new_year results in a value < 1 or > 9999 OR
+        # 1) An out-of-range year, if new_year results in a value < 1 or > 9999 - in this
+        #    case the ValueError is raised; OR
         # 2) A new month which has no corresponding day; in this case we roll forward to the
         #    first of the next month. For a defence of this logic see:
         #    tests/spec/date_series/ops/test_date_series_ops.py::test_add_months
 
         # As no month has more days than December we'll never need to roll forward from
         # December and so we don't need to worry about wrapping round to the next year
+        assert_valid_year(new_year)
```

Comment:
Sorry, I actually meant we could do this outside the `try` block, immediately after we construct `new_year`. That way we don't need to explain and handle two different error conditions inside the exception handler and I think that makes the whole thing easier to understand.

Reply:
Ah, I see - I've done that now